File: src/builtins/grep.ts
import { getWorkdir } from "../config.js";
import { clampInt } from "../utils/number.js";
import { spawnCollect } from "../utils/process.js";
import { basename, relativePath as workspaceRelativePath } from "../utils/path.js";
import { safePath } from "./fs.js";

export interface GrepInput {
  pattern: string;
  path?: string;
  glob?: string;
  max_results?: number;
  case_insensitive?: boolean;
}

export interface GrepMatch {
  file: string;
  line: number;
  column?: number;
  text: string;
}

export interface GrepResult {
  status: "ok" | "error";
  pattern?: string;
  matches?: GrepMatch[];
  truncated?: boolean;
  error?: string;
}

const DEFAULT_MAX_RESULTS = 50;
const MAX_RESULTS_CAP = 200;

export function normalizeGrepMaxResults(maxResults?: number): number {
  if (maxResults === undefined || !Number.isFinite(maxResults)) {
    return DEFAULT_MAX_RESULTS;
  }
  return clampInt(maxResults, 1, MAX_RESULTS_CAP);
}

function relativePath(absolutePath: string): string {
  const workdir = getWorkdir();
  return workspaceRelativePath(workdir, absolutePath) || basename(absolutePath);
}

interface RgMatchPayload {
  path?: { text?: string };
  lines?: { text?: string };
  line_number?: number;
  submatches?: Array<{ start?: number }>;
}

function parseRgJson(stdout: string, maxResults: number): { matches: GrepMatch[]; truncated: boolean } {
  const matches: GrepMatch[] = [];
  for (const line of stdout.split("\n")) {
    if (!line.trim()) {
      continue;
    }
    let parsed: { type?: string; data?: RgMatchPayload };
    try {
      parsed = JSON.parse(line) as { type?: string; data?: RgMatchPayload };
    } catch {
      continue;
    }
    if (parsed.type !== "match" || !parsed.data?.path?.text) {
      continue;
    }
    const data = parsed.data;
    const text = String(data.lines?.text ?? "").replace(/\n$/, "");
    matches.push({
      file: relativePath(data.path!.text!),
      line: Number(data.line_number ?? 0),
      column: data.submatches?.[0]?.start,
      text,
    });
    if (matches.length >= maxResults) {
      return { matches, truncated: true };
    }
  }
  return { matches, truncated: false };
}

function parseGrepPlain(stdout: string, maxResults: number): { matches: GrepMatch[]; truncated: boolean } {
  const matches: GrepMatch[] = [];
  for (const line of stdout.split("\n")) {
    if (!line.trim()) {
      continue;
    }
    const match = line.match(/^([^:]+):(\d+):(.*)$/);
    if (!match) {
      continue;
    }
    matches.push({
      file: match[1]!,
      line: Number(match[2]),
      text: match[3]!,
    });
    if (matches.length >= maxResults) {
      return { matches, truncated: true };
    }
  }
  return { matches, truncated: false };
}

async function searchWithRg(
  pattern: string,
  searchPath: string,
  maxResults: number,
  glob: string | undefined,
  caseInsensitive: boolean,
): Promise<GrepResult> {
  const workdir = getWorkdir();
  const args = ["--json", "--max-count", String(maxResults), pattern, searchPath];
  if (glob) {
    args.splice(1, 0, "--glob", glob);
  }
  if (caseInsensitive) {
    args.splice(1, 0, "-i");
  }

  const { stdout, stderr, code } = await spawnCollect("rg", args, { cwd: workdir });
  if (code !== 0 && code !== 1) {
    return { status: "error", error: stderr.trim() || `rg exited with code ${String(code)}` };
  }
  const { matches, truncated } = parseRgJson(stdout, maxResults);
  return { status: "ok", pattern, matches, truncated };
}

async function searchWithGrep(
  pattern: string,
  searchPath: string,
  maxResults: number,
  caseInsensitive: boolean,
): Promise<GrepResult> {
  const workdir = getWorkdir();
  const args = ["-rn", pattern, searchPath];
  if (caseInsensitive) {
    args.unshift("-i");
  }
  const { stdout, stderr, code } = await spawnCollect("grep", args, { cwd: workdir });
  if (code !== 0 && code !== 1) {
    return { status: "error", error: stderr.trim() || `grep exited with code ${String(code)}` };
  }
  const { matches, truncated } = parseGrepPlain(stdout, maxResults);
  return { status: "ok", pattern, matches, truncated };
}

export async function runGrep(input: GrepInput): Promise<string> {
  const pattern = String(input.pattern ?? "").trim();
  if (!pattern) {
    return JSON.stringify({
      status: "error",
      error: "pattern is required",
    } satisfies GrepResult);
  }

  const maxResults = normalizeGrepMaxResults(input.max_results);
  const relative = String(input.path ?? ".").trim() || ".";
  let searchPath: string;
  try {
    searchPath = safePath(relative);
  } catch (error) {
    return JSON.stringify({
      status: "error",
      error: error instanceof Error ? error.message : String(error),
    } satisfies GrepResult);
  }

  const caseInsensitive = input.case_insensitive === true;
  const glob = input.glob ? String(input.glob) : undefined;

  try {
    const rgResult = await searchWithRg(pattern, searchPath, maxResults, glob, caseInsensitive);
    if (rgResult.status === "ok") {
      return JSON.stringify(rgResult satisfies GrepResult);
    }

    const grepResult = await searchWithGrep(pattern, searchPath, maxResults, caseInsensitive);
    return JSON.stringify(grepResult satisfies GrepResult);
  } catch (error) {
    const err = error as NodeJS.ErrnoException;
    if (err.code === "ENOENT") {
      return JSON.stringify({
        status: "error",
        error: "Neither rg nor grep is available on PATH",
      } satisfies GrepResult);
    }
    return JSON.stringify({
      status: "error",
      error: error instanceof Error ? error.message : String(error),
    } satisfies GrepResult);
  }
}

File: crates/moontide-tools/src/builtins/grep.rs
use std::path::Path;

use serde_json::json;

use crate::path_util::resolve_workspace_path;

const DEFAULT_MAX: u32 = 50;
const MAX_CAP: u32 = 200;

pub async fn run_grep(
    workdir: &Path,
    pattern: &str,
    path: Option<&str>,
    glob_filter: Option<&str>,
    max_results: Option<u32>,
    case_insensitive: bool,
) -> String {
    let pattern = pattern.trim();
    if pattern.is_empty() {
        return json!({ "status": "error", "error": "pattern is required" }).to_string();
    }

    let max = normalize_max(max_results);
    let relative = path.unwrap_or(".").trim();
    let search_path = match resolve_workspace_path(relative, workdir) {
        Ok(p) => p,
        Err(e) => return json!({ "status": "error", "error": e.to_string() }).to_string(),
    };

    let out = try_rg(
        workdir,
        pattern,
        &search_path,
        max,
        glob_filter,
        case_insensitive,
    )
    .await;
    if out.get("status") == Some(&json!("ok")) {
        return out.to_string();
    }

    try_grep(workdir, pattern, &search_path, max, case_insensitive)
        .await
        .to_string()
}

fn normalize_max(max_results: Option<u32>) -> u32 {
    max_results
        .unwrap_or(DEFAULT_MAX)
        .clamp(1, MAX_CAP)
}

async fn try_rg(
    workdir: &Path,
    pattern: &str,
    search_path: &Path,
    max: u32,
    glob_filter: Option<&str>,
    case_insensitive: bool,
) -> serde_json::Value {
    let mut cmd = tokio::process::Command::new("rg");
    cmd.arg("--json")
        .arg("--max-count")
        .arg(max.to_string())
        .current_dir(workdir);
    if case_insensitive {
        cmd.arg("-i");
    }
    if let Some(g) = glob_filter {
        cmd.arg("--glob").arg(g);
    }
    cmd.arg(pattern).arg(search_path);

    match cmd.output().await {
        Ok(output) if output.status.code().is_some_and(|c| c == 0 || c == 1) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            parse_rg_json(&stdout, max)
        }
        Ok(output) => json!({
            "status": "error",
            "error": String::from_utf8_lossy(&output.stderr).trim()
        }),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            json!({ "status": "error", "error": "rg not found" })
        }
        Err(e) => json!({ "status": "error", "error": e.to_string() }),
    }
}

fn parse_rg_json(stdout: &str, max: u32) -> serde_json::Value {
    let mut matches = Vec::new();
    for line in stdout.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let Ok(parsed) = serde_json::from_str::<serde_json::Value>(line) else {
            continue;
        };
        if parsed.get("type") != Some(&json!("match")) {
            continue;
        }
        let data = &parsed["data"];
        matches.push(json!({
            "file": data["path"]["text"].as_str().unwrap_or(""),
            "line": data["line_number"].as_u64().unwrap_or(0),
            "text": data["lines"]["text"].as_str().unwrap_or("").trim_end()
        }));
        if matches.len() as u32 >= max {
            return json!({
                "status": "ok",
                "matches": matches,
                "truncated": true,
                "hint": "narrow pattern/path or lower max_results"
            });
        }
    }
    json!({
        "status": "ok",
        "matches": matches,
        "truncated": false
    })
}

async fn try_grep(
    workdir: &Path,
    pattern: &str,
    search_path: &Path,
    max: u32,
    case_insensitive: bool,
) -> serde_json::Value {
    let mut cmd = tokio::process::Command::new("grep");
    cmd.arg("-rn").current_dir(workdir);
    if case_insensitive {
        cmd.arg("-i");
    }
    cmd.arg(pattern).arg(search_path);

    match cmd.output().await {
        Ok(output) if output.status.code().is_some_and(|c| c == 0 || c == 1) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let mut matches = Vec::new();
            for line in stdout.lines() {
                if let Some((file, rest)) = line.split_once(':') {
                    if let Some((line_no, text)) = rest.split_once(':') {
                        matches.push(json!({ "file": file, "line": line_no, "text": text }));
                        if matches.len() as u32 >= max {
                            return json!({
                                "status": "ok",
                                "matches": matches,
                                "truncated": true,
                                "hint": "narrow pattern/path or lower max_results"
                            });
                        }
                    }
                }
            }
            json!({ "status": "ok", "matches": matches, "truncated": false })
        }
        Ok(output) => json!({
            "status": "error",
            "error": String::from_utf8_lossy(&output.stderr).trim()
        }),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => json!({
            "status": "error",
            "error": "Neither rg nor grep is available on PATH"
        }),
        Err(e) => json!({ "status": "error", "error": e.to_string() }),
    }
}
